### DISCUSSION TITLE
SIP must be disabled to boot + Wireless and BT discussion

### DISCUSSION THREAD

**Discussion by acermez** [AndresGarciaSobrado91/MSI-Modern15-Hackintosh#16] [General]
Hi,
Thank you so much for this repo, I bought my MSI laptop based on the findings of this repo, I'm writing this using my Hackintosh under Big Sur 11.5.2 !! I own a Z390 desktop with a nice Hackintosh setup, this is my fist laptop setup (but I'm used to the concept of kext, OC and general hackintosh debugging)

I wanted to ask about SIP, I had to use "csrutil disable" from recovery in order to boot, maybe it was a default obvious thing for this laptop setup (not for my desktop setup, working fine with SIP enabled, same OS version).

One more thing about Bluetooth, it is working with Airpods Pro, but not with my MX Master 3 (mac version) mouse, I hope this help anyone planning on using that mouse with this laptop.

I think i'm planning on buying a compatible replacement for wifi and bluetooth "Broadcom WIFI Wireless AC BCM94352Z", did anyone tried that model with this MSI laptop?

Thanks again for the awesome work :)

**Comment by AndresGarciaSobrado91** [comment]
Hi @acermez! I am pretty glad that this repo was useful for you. Regarding SIP, I'm using it enabled > `csr-active-config = 00000000`
I also have an MX Master 3 and it is not working with my Bluetooth card, but all other Bluetooth devices are working great (very strange). I have a new Broadcom wifi card (can´t remember the model). I'll test it in a few days and will update you with that information.

Have a nice week!

**Reply by acermez** [reply]
@AndresGarciaSobrado91 
Thank you for your reply, I appreciate it! 
I will try to hunt down the SIP issue and get back to you (meanwhile, I'm attaching the error screenshot, maybe someone is familiar with it, it's the "authenticating extension failed: bad code signature" part where it get stuck, endless loop of that error!). I will wait for your new wifi card test results, good luck!
Thank you.

![248326280_264544032268302_2733010554117285454_n](https://user-images.githubusercontent.com/2309533/138872834-4a738972-9bf0-4c01-9f0c-92fc3b646791.jpg)

**Comment by AndresGarciaSobrado91** [comment]
Hi! this weekend I have replaced the intel wifi card with a BCM94360NG. Sadly, I can't get it to work. I have tried several ways but it was not possible. Since I couldn't get it to work on other laptops, I assume it has a factory failure (bad for me because I bought it from China so it's hard to request a refund)

**Reply by acermez** [reply]
Damn! sorry about that! I know how it feels to order a faulty part from China, been there before :( (I live in Tunisia, 1-2 months of waiting for goods to arrive and no way to return for refund) ... I was looking at that exact same model (under Aliexpress, I assume you got it from there or Ebay). I think you choose the right model as I read all over the web that it is better than BCM94352Z. I'm traveling right now, but as soon as I get one, I will test it and let you know. I really appreciate your feedback, I hope you can get a refund or a replacement.

**Reply by acermez** [reply]
Hey again, so I ordered 2 X BCM94360NG modules, installed one and it didn't work, about to try the second one but I think I will end up with the same results... looks like there is a limitation in the bios, doing some research I found out a way to unlock the bios full features by doing (copying the following from reddit.com ..."Hold ALT, Hold Right Control, Hold Shift, And Press F2 while holding those 3 buttons", in my case since the keyboard is AZERTY, I had to add "Fn" key to the mix... I found out some advanced wireless settings such as intel's CNVI, tried changing it to disable etc... no luck, the system can't see it at the BIOS level (going to system info from bios, it says N/A for Wireless card). I hope someone can help pointing me to right way to use that full bios access to make this work. I will also try to contact MSI and see if they actually have a whitelist only list of wireless cards allowed. Thank you!

**Reply by acermez** [reply]
Quick update about my findings, the second BCM94360NG module did not work as expected (even after updating the BIOS and EC to the latest from MSI website (BIOS to E1551IMS.10F and EC Firmware to 1551EMS1.107) while disabling CNVI in hidden bios menu. The Good news is MacOs is still working as expected after the updates and the last solution I have to try before giving up is to order an M.2 E to M key adapter and loose one NVME SSD slot (goodbye dual boot on separate drives). I will keep you posted... Thanks again!

**Reply by acermez** [reply]
I tried the BCM94352Z model, same results, it didn't work... I ordered a converter from E key to M key, I will try it as soon as I get it, will keep you posted.

**Reply by AndresGarciaSobrado91** [reply]
Hi @acermez! I didn't have much time to reply to all questions on github. Thank you for your updates. I gave up with this wifi card, but if you find a way to set it up and running it would be awesome. In the next few days, I will be updating the repo to fully support macOS Monterrey. Thank you !!

**Reply by flvoicu** [reply]
Hi, I bought a bcm4350zae.
First it didn’t work because the M.2 port on the laptop is a CNViO (propietary for intel), so I bough an adapter for a normal M.2 slot.
Wifi works without any kexts in Monterrey 12.2, but I havent figured out the bluetooth part.  :/

**Reply by flvoicu** [reply]
Do you have any configurations of the wifi card in the EFI forder apart of the kexts?

**Reply by acermez** [reply]
@Flavius-Voicu That is great news, I'm really happy for you!! I tried the adapter (M.2 normal slot) but it didn't work for both models ( BCM94352Z and BCM94360NG) , I made one mistake when testing (I didn't remove the CNVI card, I will try again. Meanwhile, can you please describe the steps you did to make it work with the normal NVME M.2 port, I have the converter (I will drop a picture of it). For Bluetooth, I think we need to solder the USB cable to the USB hub or IC, not sure exactly but that's the only solution I can see happening...

![IMG_4588](https://user-images.githubusercontent.com/2309533/153698289-90e9694e-2d6a-4776-9e6f-7344681ac8fd.jpg)

**Reply by acermez** [reply]
> Do you have any configurations of the wifi card in the EFI forder apart of the kexts?

I'm not sure what do you mean :(

**Reply by acermez** [reply]
ok, great news, the M.2 adapter is working ok with BCM94360NG, I can see the card under MacOs and Windows. I will try to find a way to solder the USB for bluetooth. Thanks again @Flavius-Voicu for prompting me to test again 👍

**Reply by flvoicu** [reply]
> > Do you have any configurations of the wifi card in the EFI forder apart of the kexts?
> 
> I'm not sure what do you mean :(

As bluetooth isn’t working I thought the EFI had something more related to wifi/bluetooth apart from the kexts, but it seems I need to remap the usb ports to get bluetooth working, idk how to do it tho

**Reply by flvoicu** [reply]
> ok, great news, the M.2 adapter is working ok with BCM94360NG, I can see the card under MacOs and Windows. I will try to find a way to solder the USB for bluetooth. Thanks again @Flavius-Voicu for prompting me to test again 👍

Nice, hopfully you can figure out the bluetooth part 🤞🏻

**Reply by acermez** [reply]
> > > Do you have any configurations of the wifi card in the EFI forder apart of the kexts?
> > 
> > 
> > I'm not sure what do you mean :(
> 
> As bluetooth isn’t working I thought the EFI had something more related to wifi/bluetooth apart from the kexts, but it seems I need to remap the usb ports to get bluetooth working, idk how to do it tho

Ok I see, now I'm not sure if the intel bluetooth is not working at all for you, but in my case, it is "kind of" working, I can see devices but I can't connect to most of the devices such as my keyboard and mouse (logitech), Airpods Pro is a hit and miss, you can say it is useless, now I happen to have a USB dongle that I used before with my older desktop hackintosh (BT v4 from GMYLE, BCM20xxx), I removed the intel kexts and booted up while using the USB dongle, it did work but same unstable results, that's my findings so far, I checked for how to solder usb internally, no clean way so far, and I will have to sacrifice one USB port. My next test should be wiring the M.2 adapter BT USB to USB A-Type socket and test, if the solution is stable, I can seriously think about soldering internally... will keep you posted

**Reply by flvoicu** [reply]
Well I have the same intel bluetooth issues so I am trying to downgrade to big sur 11.5.2 to see if that will fix it. 
If you want to try the brcm bluetooth USB there are 9pins adaptors that plug to a normal USB,   https://www.amazon.com/dp/B07Q4QZW57.
If downgrading does not get the job done I will buy one and maybe I am able to open it and cut the excess metal of the port so it does not pertrude too much.

**Reply by flvoicu** [reply]
Update, bluetooth and continuity features working in big sur 11.5.2, so I am staying until it gets fixed in monterey ;)

**Comment by acermez** [comment]
@Flavius-Voicu , Nice! I'm glad you figured it out, and so sorry I'm late to reply here, I didn't notice your first reply about downgrading and we didn't mention our MacOS versions before so I assumed you are under Big Sur already (I'm using 11.6.2), as for the normal USB adapter, thank you but it would be faster for me to solder the 4 wires to an old USB male socket (I have some laying around, after all its +5v, data+, data- and GND , so soldering each wire to the same color should do the job. I'm sure I will get the same results as you (working wifi and BT with native behaviour). I might do the same and cut a small hole for the wire to be inserted in one of the USB ports, still not sure, I really like the soldering idea, looks cleaner to me, I will continue to post my progress!

**Reply by flvoicu** [reply]
Np for the reply, btw I would recommend to downgrade to 11.5.2 and use intel bluetooth and brcm wifi, so you dont loose a port until the intel bluetooth gets fixed in monterrey or big sur 11.6.2.

**Reply by acermez** [reply]
I really tried to make this cleaner by soldering the USB data + / data - to the M.2 Wifi (E Key) socket, all the documentation I was able to find pointed that pin 3 and 5 are for BT USB, it didn't work, I hope someone can figure this one out someday, I tried all I can do from the BIOS, nothing... it made me really frustrated I went with passing the wires from a small gap under the SD Card Reader so it's a non destructive solution and connecting the USB to the nearest USB A port. I could solder it inside or even solder it to the SD Card USB lanes but I don't think it is worth it anymore... I'm dropping a couple pics of my ugly/working solution.
![IMG_4621](https://user-images.githubusercontent.com/2309533/154791201-6258082a-1501-4306-8e10-64bbf5c33683.jpg)
![IMG_4623](https://user-images.githubusercontent.com/2309533/154791202-f4fb44dc-9c55-46dc-9d80-4245e2365ab3.jpg)
![IMG_4624](https://user-images.githubusercontent.com/2309533/154791204-0168068d-5d15-4556-95e6-0867f32f2535.jpg)
![IMG_4625](https://user-images.githubusercontent.com/2309533/154791207-621025dd-d09a-4501-9ce8-3012c67a7a55.jpeg)
![IMG_4626](https://user-images.githubusercontent.com/2309533/154791209-a525447d-8094-410c-b02b-25297d3ad272.jpg)

**Comment by acermez** [comment]
Hey there, last update about this... the above ugly solution was an obvious reply to the stupid intel's CNVI limitations. Now that the BCM card did work as expected, I removed everything and I was able to keep the Intel wifi with a BCM BT USB dongle I had laying around since my early hackintosh days... I'm sharing the pics of the dongle just in case anyone wants to order it. Thank you!
![IMG_4635](https://user-images.githubusercontent.com/2309533/155563570-f7d30921-0e20-49b4-a6cf-1cb468b60b88.jpeg)
![IMG_4636](https://user-images.githubusercontent.com/2309533/155563591-5c7a935e-76b2-4360-986a-8dde1de7ed18.jpeg)
